components/layout/display_list/builder.rs

```diff
@@ -1319,110 +1330,102 @@ impl FragmentDisplayListBuilding for Fragment {
+                    border_widths,
+                    border_style_struct,
+                    |fallback_size| match image {
+                        Image::Url(ref image_url) => image_url
+                            .url()
+                            .and_then(|url| {
+                                state.layout_context.get_webrender_image_for_url(
+                                    self.node,
+                                    url.clone(),
+                                    UsePlaceholder::No,
+                                )
+                            }).and_then(convert_image_to_border),
+                        Image::PaintWorklet(ref paint_worklet) => self
+                            .get_webrender_image_for_paint_worklet(
+                                state,
+                                style,
+                                paint_worklet,
+                                fallback_size,
+                            ).and_then(convert_image_to_border),
+                        Image::Gradient(ref gradient) => match gradient.kind {
+                            GradientKind::Linear(angle_or_corner) => {
+                                let (wr_gradient, linear_stops) = convert_linear_gradient(
+                                    style,
+                                    fallback_size,
+                                    &gradient.items[..],
+                                    angle_or_corner,
+                                    gradient.repeating,
+                                );
+                                stops = linear_stops;
+                                Some((
+                                    NinePatchBorderSource::Gradient(wr_gradient),
+                                    convert_size_au_to_u32(fallback_size),
+                                ))
+                            },
+                            GradientKind::Radial(shape, center, _angle) => {
+                                let (wr_gradient, radial_stops) = convert_radial_gradient(
+                                    style,
+                                    fallback_size,
+                                    &gradient.items[..],
+                                    shape,
+                                    center,
+                                    gradient.repeating,
+                                );
+                                stops = radial_stops;
+                                Some((
+                                    NinePatchBorderSource::RadialGradient(wr_gradient),
+                                    convert_size_au_to_u32(fallback_size),
+                                ))
+                            },
+                        },
+                        _ => None,
+                    },
+                ).map(|(details, border)| {
+                    // The image's border width can differ from the "simple" color border width.
+                    layout_border_width = border;
```

Comment:
This feels a bit out of place, but I don't have immediately better suggestions...